### PR TITLE
add validate to components

### PR DESCRIPTION
**Pull request by glmnet**
# What does this implement/fix? 

A mechanism for components to validate the configuration after all components passed schema validation.

Initially just parent component validation was attempted, however a more generic last time validation can be more useful.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Configuration change (this will require users to update their yaml configuration files to keep working)

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>
  
# Test Environment

- [ ] ESP32
- [ ] ESP8266
- [x] Windows
- [ ] Mac OS
- [ ] Linux

## Example entry for `config.yaml`:

image

# Explain your changes

This sends the config object to the component and the config of the item in the optional validate module funciton:

````python
def validate(config, item_config):
# some components like spi, uart, can have validation options, e.g. sim800l requires uart to be 9600 baud
  uart.validate_device("sim800l", config, item_config, baud_rate=9600)

````

# Also added some helper methods in config to deal with getting configuration by id:  
# eg. in `uart.validate_device` the parent uart configuration is needed: 

````python
  uart_config = config.get_config_by_id(item_config[CONF_UART_ID])
````

Useful in the given example, where spi component is okay to not declare both MISO and MOSI but `rc522_spi` (and many others) do require it to be bidirectional comm.

Other ideas are, e.g. components checking 
- uart config is ok (baud rate, stop bits)
- an output supports PWM e.g. for `rtttl` 
- the duty cycle is not 100% on rc transmit IR

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by glmnet** [comment]
cc @SenexCrenshaw

**Comment by SenexCrenshaw** [comment]
Woot! Looking forward to adding this into spi and my other devices. This is great and thank you!

**Comment by jesserockz** [comment]
Wouldn't this cause false errors if you had multiple "parents" for one component? Because there is no way to specify that `def validate_parent(property, config, component):` should only apply to (in this case) the SPI component. 

I can't think of any examples off the top of my head, but I am sure there could be some...

**Comment by glmnet** [comment]
> Wouldn't this cause false errors if you had multiple "parents" for one component?

Did not test that, but it shouldn't really.

I'll take the opportunity to better document the function.

After all validations are done, ESPHome will call to `validate_parent` for each config element which has a `cv.use_id` in your configuration:
* `property`: this contains the name of the property with the `cv.use_id`, e.g. the `rc522_spi` itself does not have a `cv.use_id` but the inherited spi schema does, and the name of the property is `spi_id` so that is what you get there. The id itself in this case is autogenerated as in the yaml example is not used, this id is not passed in as it is really not useful. 
* `config`: this is the validated section of the CONFIG of the *resolved parent* component. *The way the code is written retrieves the config based on the path of the associated ID*
* `component`: This is the ComponentManifest (an object that get's you access to the `.py` module of, in this case, the SPI component) I dont know how useful can be as didn't try to make many parent validators yet.

**Comment by jesserockz** [comment]
I guess what I mean to say is if you have multiple "parents" you need to write the `validate_parent` function so it only checks certain config keys based on the `property`.

From this example it is hard to know that, but yeah documentation, or more examples (or this one) that check the `property` before checking the config keys

**Comment by mmakaay** [comment]
Very cool. Based on the description, this provides exactly the features that I was looking for:

* a final validation step in which I can check a component config against the full config (to make sure the component config uses valid references to elements configured by another component) 
* and cleanly retrieving component configuration by name.

I'll go ahead and test drive this.

**Comment by jesserockz** [comment]
Just be careful with naming as `def validate()` exists in some components already as a config validator.

**Comment by mmakaay** [comment]
Yes, the naming might be a bit generic. Would it be an idea to stick close to the naming of `CONFIG_SCHEMA`, to have it stand out as a framwork construct? E.g.

```python
def CONFIG_VALIDATOR(config, component_config):
    # do
    # your
    # thing
```

It might be very useful to be able to still tweak the component config from this validation function, in the same way as it is possible with validators in voluptuous (by returning the updated config). This can be useful to fill in default values for optional parameters, which do depend on settings in or availability of other components.

Returning an updated component config could be optional. The calling code would apply the returned component config to the global config,  unless it is `None`.

**Comment by mmakaay** [comment]
In the implementation code, I see that the following constraint is implemented in the validation loop:

```python
            if comp.config_schema is None:
                continue
```

Is that check really needed? I would think that a component without a schema, but with a final validator function would be fine. It might not be the most prevalent use case, but I don't think it's required to punish this kind of implementation by  not running the final validator.

**Comment by glmnet** [comment]
> Just be careful with naming as `def validate()` exists in some components already as a config validator.

I've just renamed all components which use `validate` to `_validate`, since that validate was always private.

So test are passing now (otherwise it was calling with different set of params) also had to modify some tests as now validation is more strict e.g. in uart not being shared.

Do you think it will better to have another name instead of `validate`?

**Comment by glmnet** [comment]
> In the implementation code, I see that the following constraint is implemented in the validation loop:
> 
> ```python
>             if comp.config_schema is None:
>                 continue
> ```
> 
> Is that check really needed? I would think that a component without a schema, but with a final validator function would be fine. It might not be the most prevalent use case, but I don't think it's required to punish this kind of implementation by not running the final validator.

Just because it was already there and the use case hasn't showed up yet. Can be added when it's proven to be needed.

**Comment by glmnet** [comment]
> ```python
> def CONFIG_VALIDATOR
> ```

I guess ALL_UPPERCASE are for constant stuff? may be is just convention, but there is also the `def to_code()` which is lower case function, so merging with upper case function looks messy to me.

**Comment by mmakaay** [comment]
Agreed on all points. 
With the 'validate' conflicts out of the way, the search for alternatives is no longer needed. 
Checking for a config schema adds a bit of coupling, but practically this does not and unlikely will bite me.

File: tests/test3.yaml [hunk 1]
```diff
@@ -408,32 +433,16 @@ sensor:
...
-  - platform: pmsx003
-    type: PMS5003ST
-    pm_2_5:
-      name: 'PM 2.5 Concentration'
-    temperature:
-      name: 'PMS Temperature'
-    humidity:
-      name: 'PMS Humidity'
-    formaldehyde:
-      name: 'PMS Formaldehyde Concentration'
```
Comment (jesserockz) [inline tests/test3.yaml:435]: @glmnet Why did you remove these from the test?

**Review by jesserockz — APPROVED** [review]
Just one question about the test file above. Otherwise, lets get this merged.

File: tests/test3.yaml (hunk 1 above)
Comment (glmnet) [reply tests/test3.yaml:435]: Ahh this is ok, removed because I would need to add more "`uarts`" so validation will pass, and the objective of having code being compiled is already done by the inclusion of one of these sensors just before the other two types I'm removing, i.e. the only difference of having several types deads in `cg.add(var.set_type(config[CONF_TYPE]))`
